### PR TITLE
Complete Resorcerer fixed-point field type support

### PR DESCRIPTION
**Pull request by CelticMinstrel**
When I was looking at your list of supported field types, it occurred to me that most of the unsupported ones should be pretty simple to add support for, especially the real-number ones.

So, I whipped up some code to add support for the following Resorcerer field types:

* `FRAC`, `SFRC`, `FWID`, and `FXYZ`, all of which are just minor variations on `FIXD`. Maybe they could all be handled as the same class somehow but I didn't look into that.
* ~`DBDB`, which is literally just a pair of doubles. I modelled it off `PNT ` but also added the formatter override from `DOUB`.~ (Removed by request)
* ~`EXTN`, which is the Swift `Float80` type. Since `Float80` doesn't have `.bitPattern`, I had to write code to read/write the exponent and mantissa separately.~ (Removed by request)
* ~`XT96` and `UNIV`, which are the same as `EXTN` but with 2 bytes of padding added.~ (Removed by request)

Unfortunately, it turns out I'm not able to build ResForge myself, because I'm on MacOS 11 and Xcode 12. I did verify that my code doesn't introduce any new compiler errors (by telling Xcode to build only the Template Editor and littering the code with `!`'s and `?`'s to satisfy the older compiler), but I can't be sure it actually works. Even so, I thought it was worth bringing it to your attention.

I have a spare old Mac that can run the OSX version of Resorcerer, so if needed I could provide a .rsrc file with a template and sample resource that uses these types for testing purposes. Alternatively, I could test it myself if I just had a working build (the released build runs on my machine; I just can't build it myself).

### CODE REVIEW

**Comment by andrews05** [comment]
Hi @CelticMinstrel, thanks for the PR!

I have to ask, do you actually have a use for these types? The reason I never implemented them was because I couldn't find any templates that used them (possibly except for some that required a FLTR, which ResForge can't support anyway).

I'm not totally opposed to including them, though it would be good to have a reason to do so. It wasn't my objective to reach feature parity with Resorcerer, especially as there are some things I don't like about it 😄

**Review by CelticMinstrel — COMMENTED** [review]
Since you mentioned it, I looked through all of Resorcerer's templates and indeed couldn't find any uses of them. Since `FWID` stands for "Font Width", I thought maybe some of the font templates would use them (there's even an `FWID` template!), but nope. (Maybe `sfnt` or something uses it; I skipped any that need filters.)

I think the `PICT` template uses `SFRC`, but again that one is a filtered template.

I also scoured through old templates I had laying around from years ago (most made for Rezilla or ResEdit, rather than Resorcerer; but Rezilla does support all of these) and didn't find any uses of them.

My primary motivation for doing this wasn't because I specifically needed them but because it seemed like there was no reason to _not_ support them – the work to support these is _almost_ the same as the work to support `REAL`, `DOUB`, and `FIXD`. You can never know when someone might show up with ancient templates that need these types, and if that ResForge already supports them, they can be saved the trouble of trying to somehow run Resorcerer (or Rezilla) to access the data (or extracting it by other, inconvenient means). I am not that person, and I don't know if that person even exists. If it was tricky to support these, that would be one thing; but the additional fixed-point types are no more complicated than `FIXD`, and `EXTN` is only slightly more complicated than `DOUB`. And `UNIV` / `XT96` is only slightly more complicated than `EXTN`.

Of course, I wouldn't expect you to reach feature parity with Resorcerer, as their filter system is pretty hard to replicate, and actually directly supporting their filter system would mean embedding an M68k emulator into your program, which is overkill for what is mostly a niche use-case (I think). Similarly, the +/- types are very niche, seemingly intended as a way to migrate old resources when a template format is changed. But I also think there is benefit in at least clearing all the low-hanging fruit.

(I am curious as to what you dislike about Resorcerer. I can't say I especially like it myself, but it does have at least one useful thing that ResForge lacks, which is the ability to insert anywhere in a list rather than only at the end. Certainly it was a major step up from ResEdit, but that's not saying much now. One thing I miss from ResEdit though was that every resource had a nice 32x32 colour icon… none of the other resource editors replicate that, though ResForge does at least have an icon.)

File: Plugins/Sources/TemplateEditor/Elements/ElementFRAC.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    override func writeData(to writer: BinaryDataWriter) {
+        writer.write(intValue)
+    }
+
+    override var formatter: Formatter {
+        self.sharedFormatter {
+            let formatter = NumberFormatter()
+            formatter.hasThousandSeparators = false
+            formatter.numberStyle = .scientific
+            formatter.maximumFractionDigits = 5
```
Comment (andrews05) [inline Plugins/Sources/TemplateEditor/Elements/ElementFRAC.swift]: All the fixed fractions should work fine as decimal. Since FRAC has 30 fractional bits, we can push the digits out to 9.
```suggestion
            formatter.numberStyle = .decimal
            formatter.maximumFractionDigits = 9
```

File: Plugins/Sources/TemplateEditor/Elements/ElementFWID.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    override func writeData(to writer: BinaryDataWriter) {
+        writer.write(intValue)
+    }
+
+    override var formatter: Formatter {
+        self.sharedFormatter {
+            let formatter = NumberFormatter()
+            formatter.hasThousandSeparators = false
+            formatter.numberStyle = .decimal
+            formatter.maximumFractionDigits = 5
```
Comment (andrews05) [inline Plugins/Sources/TemplateEditor/Elements/ElementFWID.swift]: Only 12-bits here, so 4 digits might be better.
```suggestion
            formatter.maximumFractionDigits = 4
```

File: Plugins/Sources/TemplateEditor/Elements/ElementSFRC.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        writer.write(intValue)
+    }
+
+    override var formatter: Formatter {
+        self.sharedFormatter {
+            let formatter = NumberFormatter()
+            formatter.hasThousandSeparators = false
+            formatter.numberStyle = .scientific
+            formatter.maximumFractionDigits = 5
+            formatter.minimum = Double(UInt16.min) / Self.fixed1 as NSNumber
```
Comment (andrews05) [inline Plugins/Sources/TemplateEditor/Elements/ElementSFRC.swift]: Decimal again here. Since this is unsigned, we can just set minimum to 0.
```suggestion
            formatter.numberStyle = .decimal
            formatter.maximumFractionDigits = 5
            formatter.minimum = 0
```

**Review by andrews05 — CHANGES_REQUESTED** [review]
Having reviewed the code, I'd be happy to accept the fixed fractions with a few small tweaks.

I'd prefer to skip the others for now. `Float80` is actually hardware-dependent - it's not supported on arm so we can't use it here. `DBDB` doesn't quite work with RECT's configure but it doesn't add anything of value anyway since you can just do `DOUB`, `DOUB`.

**Comment by andrews05** [comment]
> I am curious as to what you dislike about Resorcerer.

Mostly I found the UI incredibly cumbersome, particularly not being able to edit values directly in the template (it had to open a new window just to edit a field).
But I also found the design of some fields and behaviours rather odd. Such as:
- Two variants of endian change fields (the visible ones are unnecessary - you could add a comment field if you needed to show something)
- The +/- types you mentioned (ResForge could easily support these but it seems both bizarre and dangerous - surely you would have a better way to migrate all your data to a new format?)
- The LSTS/HEXS fields, which are redundant (ResForge allows LSTB/HEXD to be used in the same manner)

> I can't say I especially like it myself, but it does have at least one useful thing that ResForge lacks, which is the ability to insert anywhere in a list rather than only at the end.

Ah, I'll admit it's not exactly intuitive by today's standards but ResForge actually works the same as ResEdit for this: Press cmd+K to insert an item before the currently selected one. (You can also cut/copy/paste list items)

> One thing I miss from ResEdit though was that every resource had a nice 32x32 colour icon… none of the other resource editors replicate that, though ResForge does at least have an icon.

Indeed, I certainly liked the icons too. In ResForge the icons for the EV Nova types are actually all emojis 😆

**Review by CelticMinstrel — COMMENTED** [review]
I guess if ARM doesn't support `Float80` at all then there's not much to be done about that (you could pull in a software floating-point emulation library, but that no longer falls in the realm of "so easy there's no reason not to").

> Mostly I found the UI incredibly cumbersome, particularly not being able to edit values directly in the template (it had to open a new window just to edit a field).

Ah yes, that _is_ quite cumbersome! And for compound types like Rect you still only have one edit field. And there's no bounds checking, so you can happily enter 99999999999 in a `FIXD` and it'll accept it, silently clamping it when you save the resource.

Anyway, I'll make the requested changes soon. Feel free o use the squash button to merge everything into one commit.

**Review by CelticMinstrel — COMMENTED** [review]
>  `DBDB` doesn't quite work with RECT's configure but it doesn't add anything of value anyway since you can just do `DOUB`, `DOUB`.

In my opinion, supporting templates that use them without any additional work does count as something of value, though I'll admit it's not that high of a value when the person could edit their template and get it working.

I've made the requested changes and also updated the README to mark those types as supported (though I didn't update the count at the top of the file, as I'm a bit unsure what it's actually counting). I'm also marking it ready for review.

File: Plugins/Sources/TemplateEditor/Elements/ElementFWID.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    override var formatter: Formatter {
+        self.sharedFormatter {
+            let formatter = NumberFormatter()
+            formatter.hasThousandSeparators = false
+            formatter.numberStyle = .decimal
+            formatter.maximumFractionDigits = 4
+            formatter.minimum = Double(Int16.min) / Self.fixed1 as NSNumber
+            formatter.maximum = Double(Int16.max) / Self.fixed1 as NSNumber
```
Comment (andrews05) [inline Plugins/Sources/TemplateEditor/Elements/ElementFWID.swift:33]: So I had test of the new types and found an annoying edge case. For FWID and FXYZ only, the max value will be rounded up for display. That rounded value is now _above_ the max allowed by the formatter and it won't accept it again if you try to edit it. Rather than try complicated ways to calculate the max _display_ value, it's probably best to just set them directly.
```suggestion
            // The actual max value will be rounded up for display, so we should set
            // the min/max display values directly instead of trying to calculate them.
            formatter.minimum = -8
            formatter.maximum = 7.9998
```

File: Plugins/Sources/TemplateEditor/Elements/ElementFXYZ.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    override var formatter: Formatter {
+        self.sharedFormatter {
+            let formatter = NumberFormatter()
+            formatter.hasThousandSeparators = false
+            formatter.numberStyle = .decimal
+            formatter.maximumFractionDigits = 5
+            formatter.minimum = Double(Int16.min) / Self.fixed1 as NSNumber
+            formatter.maximum = Double(Int16.max) / Self.fixed1 as NSNumber
```
Comment (andrews05) [inline Plugins/Sources/TemplateEditor/Elements/ElementFXYZ.swift:33]:
```suggestion
            formatter.minimum = -1
            formatter.maximum = 0.99997
```

**Review by andrews05 — CHANGES_REQUESTED** [review]
Thanks for the changes, it looks pretty good!

The count in the read me is the number of types we support, so you can change the 64 to 68.

**Review by CelticMinstrel — COMMENTED** [review]
Made those changes, also replicating your comment in FWID to FXYZ.

**Review by andrews05 — APPROVED** [review]
Nice, thanks again!

We can always look at other fields in future when and if anybody needs them.

**Review by CelticMinstrel — COMMENTED** [review]
Just a random note, slightly relevant but mostly tangential: I just noticed that the `snd ` resource format uses the `EXTN` type. It would probably be difficult to make a template for `snd ` though, plus you already have a dedicated editor. But that does show it's used _somewhere_.

If ARM doesn't support `Float80`, I expect an implementation would need to convert it to either `Float128` (if that exists and is supported) or `Float64` (aka `Double`).

I stashed my implementation of `EXTN` just in case it's useful in the future, not that I really expect it to be.
